db/src/cache.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::BTreeMap;
+use std::cmp::Ord;
+use std::fmt::Debug;
+
+use rusqlite;
+
+use errors::{
+    Result
+};
+use db::{
+    TypedSQLValue,
+};
+use mentat_core::{
+    Entid,
+    TypedValue,
+};
+
+pub type CacheMap<K, V> = BTreeMap<K, V>;
+
+pub trait ValueProvider<K, V>: Sized {
+    fn fetch_values<'sqlite>(&mut self, sqlite: &'sqlite rusqlite::Connection) -> Result<CacheMap<K, V>>;
+}
+
+pub trait Cacheable {
+    type Key;
+    type Value;
+    type ValueProvider;
```

Comment:
Shouldn't this be eliminated and `ValueProvider<Self::Key, Self::Value>` used in `new`?